src/tentaclio/clients/google_drive_client.py

```diff
@@ -0,0 +1,109 @@
+"""Local filesystem client."""
+import json
+import os
+from typing import Any, Iterable, Optional, Tuple, Union
+
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+
+from tentaclio import fs, protocols, urls
+
+from . import base_client
+
+
+__all__ = ["GoogleDriveFSClient"]
+
+# Load the location of the token file from the environment
+TOKEN_FILE = os.getenv(
+    "TENTACLIO__GOOGLE_DRIVE_TOKEN_FILE", os.environ["HOME"] + os.sep + ".google_drive_token.json",
+)
+
+
+def _load_credentials(token_file: str) -> Credentials:
+    """Load the credentials and refresh them if necesary."""
+    creds = None
+    if os.path.exists(token_file):
+        with open(token_file) as f:
+            state = json.load(f)
+            creds = Credentials(**state)
+    else:
+        raise ValueError(f"Token file is not valid {token_file}")
+
+    # If there are no (valid) credentials available refresh them or raise an error.
+    if not creds or not creds.valid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            raise ValueError(f"Couldn't refresh token in f{token_file}")
+        # Save the credentials for the next run
+        with open(token_file, "w") as f:
+            f.write(creds.to_json())
+    return creds
+
+
+class GoogleDriveFSClient(base_client.BaseClient["GoogleDriveFSClient"]):
+    """Allow filesystem-like access to google drive."""
+
+    DEFAULT_DRIVE = "root"
+
+    allowed_schemes = ["gdrive", "googledrive"]
+
+    drive: str
+    parts: Tuple[str]
+
+    # Not an easy task to figure out the type of the
+    # returned value from the library
+    _service: Optional[Any] = None
+
+    def __init__(self, url: Union[urls.URL, str]) -> None:
+        """Create a new GoogleDriveFSClient."""
+        super().__init__(url)
+
+        if len(self.url.path) == 0:
+            raise ValueError(
+                f"Bad url: {self.url.path} .Google Drive needs at least "
+                "the drive part (i.e. gdrive://My Drive)"
+            )
+        parts = self.url.path.split("/")
+        # FIXME for the moment the drive is pointing to the magic root
+        # drive (My Drive)
```

Comment:
don't worry about this too much I will make it cleaner when I need to fight against the _drive_ model google drive follows